tests/cpp/test_gpu_view.cpp

```diff
@@ -2344,4 +2344,128 @@ TEST_F(GpuViewTest, ExpandedBroadcast) {
+  auto t0 = at::randn(input_shape, options);
+  auto tw = at::randn(input_shape_wb, options_wb);
+  auto tb = at::randn(input_shape_wb, options_wb);
+  auto t1 = t0.reshape(group_shape).to(at::kFloat);
+  auto t2 = t1.sum({-1, -2, -3}).unsqueeze(-1).unsqueeze(-1).unsqueeze(-1);
+  auto t3 = t1 / t2;
+  auto t4 = t3.reshape(input_shape);
+  auto t5 = tw.unsqueeze(0).unsqueeze(-1).unsqueeze(-1);
+  auto t6 = tb.unsqueeze(0).unsqueeze(-1).unsqueeze(-1);
+  auto t7 = t4.mul(t5).add(t6);
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+  auto cg_outputs = executor_cache.runFusionWithInputs({t0, tw, tb});
+  // should expect 1 after adding a pre-segment pass to move reshape to input
+  // and output.
+  auto seg_groups =
+      executor_cache.getMostRecentKernelRuntime()->fusionSegments()->groups();
+  int n_pointwise = 0, n_norm = 0, n_other = 0;
+  for (auto sg : seg_groups) {
+    if (sg->heuristic() == ScheduleHeuristic::PointWise) {
+      n_pointwise++;
+    } else if (sg->heuristic() == ScheduleHeuristic::InnerPersistent) {
+      n_norm++;
+    } else {
+      n_other++;
+    }
+  }
+  EXPECT_EQ(n_pointwise, 2);
+  EXPECT_EQ(n_norm, 1);
+  EXPECT_EQ(n_other, 0);
+  testValidate(
+      executor_cache.fusion(),
+      cg_outputs,
+      {t0, tw, tb},
+      {t7},
+      __LINE__,
+      __FILE__);
+}
+
+TEST_F(GpuViewTest, OutputAliasIntermediate) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  const int64_t N = 2, C = 128, H = 16, W = 16, G = 32;
+  const std::vector<int64_t> input_shape = {N, G, C / G, H, W};
+  const std::vector<int64_t> output_shape = {N, C, H, W};
+  DataType dtype = DataType::Half;
+  auto tv0 = makeContigTensor(input_shape.size(), dtype);
+  fusion->addInput(tv0);
+  auto tv1 = castOp(DataType::Float, tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = sum(tv2, {-1, -2, -3});
+  auto tv4 = broadcast(tv3, {false, false, true, true, true});
+  auto tv5 = div(tv2, tv4);
+  auto tv6 = castOp(dtype, tv5);
+  auto tv7 = reshape(tv6, input_shape, output_shape);
+  fusion->addOutput(tv7);
+
+  auto options =
+      at::TensorOptions().dtype(data_type_to_aten(dtype)).device(at::kCUDA, 0);
+  auto t0 = at::randn(input_shape, options);
+  auto t1 = t0.to(at::kFloat);
+  auto t2 = t1.sum({-1, -2, -3}).unsqueeze(-1).unsqueeze(-1).unsqueeze(-1);
+  auto t3 = t1 / t2;
+  auto t4 = t3.reshape(output_shape);
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+  auto cg_outputs = executor_cache.runFusionWithInputs({t0});
+  auto seg_groups =
+      executor_cache.getMostRecentKernelRuntime()->fusionSegments()->groups();
+  int n_no_op = 0, n_norm = 0;
+  for (auto sg : seg_groups) {
+    if (sg->heuristic() == ScheduleHeuristic::NoOp) {
+      n_no_op++;
+    } else if (sg->heuristic() == ScheduleHeuristic::InnerPersistent) {
+      n_norm++;
+    } else {
+      FAIL() << "Unexpected heuristic: " << sg->heuristic();
+    }
+  }
+  EXPECT_EQ(n_no_op, 1);
+  EXPECT_EQ(n_norm, 1);
```

Comment:
This can be simplified using gmock. Something like: 

```
EXPECT_THAT(groups, Contains(HeuristicIs(NoOp)).Times(1));
EXPECT_THAT(groups, Contains(HeuristicIs(InnerPersistent)).Times(1));
```

See https://github.com/NVIDIA/Fuser/blob/3c97a085d44a4f34c947efd198dda7b8afad3d56/tests/cpp/test_move_split_cat.cpp#L206 and https://github.com/NVIDIA/Fuser/blob/3c97a085d44a4f34c947efd198dda7b8afad3d56/tests/cpp/test_alias.cpp#L1091-L1093.

Reply:
New Knowledge Received!